### ISSUE TITLE
plot_correlation error: correlation is nan when column has nan value

### ISSUE THREAD

**Issue by eutialia** [sfu-db/dataprep#214]
**Describe the bug**
plot_correlation(df) will have nan as correlation value.
plot_correlation(df, x, y) will not work.

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
 in 
      1 from dataprep.eda import plot, plot_correlation, plot_missing
----> 2 plot_correlation(df, 'Annual Salary', 'Typical Hours')

~/Source/dataprep/dataprep/eda/correlation/__init__.py in plot_correlation(df, x, y, value_range, k)
     63     """
     64 
---> 65     intermediate = compute_correlation(df, x=x, y=y, value_range=value_range, k=k)
     66     figure = render_correlation(intermediate)
     67 

~/Source/dataprep/dataprep/eda/correlation/compute.py in compute_correlation(df, x, y, value_range, k)
    145                 df[y].values.compute_chunk_sizes(),
    146                 k=k,
--> 147                 sample_size=1000,
    148             )
    149             result = {

~/Source/dataprep/dataprep/eda/correlation/compute.py in scatter_with_regression(xarr, yarr, sample_size, k)
    196     xarrp1 = da.vstack([xarr, da.ones_like(xarr)]).T
    197     xarrp1 = xarrp1.rechunk((xarrp1.chunks[0], -1))
--> 198     (coeffa, coeffb), _, _, _ = da.linalg.lstsq(xarrp1, yarr)
    199 
    200     if sample_size < len(xarr):

~/Source/dataprep/.venv/lib/python3.7/site-packages/dask/array/linalg.py in lstsq(a, b)
   1244     """
   1245     q, r = qr(a)
-> 1246     x = solve_triangular(r, q.T.dot(b))
   1247     residuals = b - a.dot(x)
   1248     residuals = (residuals ** 2).sum(keepdims=True)

~/Source/dataprep/.venv/lib/python3.7/site-packages/dask/array/linalg.py in solve_triangular(a, b, lower)
    977     if b.ndim <= 2:
    978         if a.shape[1] != b.shape[0]:
--> 979             raise ValueError("a.shape[1] and b.shape[0] must be equal")
    980         if a.chunks[1] != b.chunks[0]:
    981             msg = (

ValueError: a.shape[1] and b.shape[0] must be equal
```

**To Reproduce**

```python
df = pd.read_csv('https://data.cityofchicago.org/api/views/xzkq-xp2w/rows.csv?accessType=DOWNLOAD')
plot_correlation(df)
plot_correlation(df, 'Annual Salary', 'Typical Hours')
```

**Expected behavior**
No nan value

**Screenshots**
<img width="330" alt="image" src="https://user-images.githubusercontent.com/12697964/86428870-a78d4a80-bca2-11ea-8a8a-a57936829de2.png">

**Desktop (please complete the following information):**
 - OS: macOS
 - Browser: Safari
 - Python Version: 3.7.7
 - Dataprep Version: 0.2.7

**Comment by jinglinpeng** [comment]
@Waterpine  Could you please take a look?

**Comment by Waterpine** [comment]
I think the reason is that plot_correlation does not deal with nan/null value.
<img width="1110" alt="屏幕快照 2020-07-09 上午12 01 43" src="https://user-images.githubusercontent.com/29000790/86942214-7667bb00-c177-11ea-8e55-1af444a0e50a.png">
Thus, we have to consider nan/null value in the plot_correlation function? We have to deal with nan/null value first, then we compute the correlation between columns? @jinglinpeng

**Comment by jinglinpeng** [comment]
@Waterpine  I think we should delete nan/null value first, which is also done by `pandas.Dataframe.corr.`

**Comment by dovahcrow** [comment]
This can be solved by #223 right?

**Comment by jinglinpeng** [comment]
> This can be solved by #223 right?

No, they are different task.

**Comment by dovahcrow** [comment]
> > This can be solved by #223 right?
> 
> No, they are different task.

I think you can call the drop_null before computing the correlations.

**Comment by jinglinpeng** [comment]
> > > This can be solved by #223 right?
> > 
> > 
> > No, they are different task.
> 
> I think you can call the drop_null before computing the correlations.

Yes. This task is to call drop_null before computing correlations, while #223 is how to compute drop_null. They are different but related.

**Comment by dovahcrow** [comment]
Another thing I want to include in the change is to create a function `fn preprocess(df)`, which not only do drops_null but also move our existing logics for detecting the dtypes and converting the pandas dataframe to dask dataframe into the preprocess function.

**Comment by jinglinpeng** [comment]
> Another thing I want to include in the change is to create a function `fn preprocess(df)`, which not only do drops_null but also move our existing logics for detecting the dtypes and converting the pandas dataframe to dask dataframe into the preprocess function.

The preprocessing part also varies based on different sub-functions (e.g., `plot(df)`, `plot(df, x)`), and hence they are processed inside each sub-function currently. For example, user may only call `plot(df, x)`, and in this case we only need to drop null and detect type for x column.

**Comment by dovahcrow** [comment]
That can be differentiated by the parameters of `preprocess`. Basically I think it is better to put preprocessing in one place so that all our functions have the same structure thus easier to maintain.


Regards,
Weiyuan Wu

---
Weiyuan Wu | PhD Student
a: SFU-DB | 8888 University Dr. | Burnaby, BC, CA
e: youngw@sfu.ca | w: wooya.me

**Comment by jinglinpeng** [comment]
> That can be differentiated by the parameters of `preprocess`. Basically I think it is better to put preprocessing in one place so that all our functions have the same structure thus easier to maintain.

IC. Agree.

**Comment by Waterpine** [comment]
I wonder whether Peter has implemented drop_null function. Or I just use df.dropna() first, then I replace it with drop_null function after Peter implements it? @dovahcrow @jinglinpeng

**Comment by jinglinpeng** [comment]
> I wonder whether Peter has implemented drop_null function? Or I just use df.dropna() first, then I replace it with drop_null function after Peter implements it? @dovahcrow @jinglinpeng

@Waterpine  Peter is working on `drop_null`. I think you can work on #243 first, and then back to this issue when `drop_null` is merged.
